File: java/src/main/java/colorfightII/Colorfight.java
package colorfightII;

import org.json.simple.JSONArray;
import org.json.simple.JSONObject;
import org.json.simple.parser.JSONParser;
import org.json.simple.parser.ParseException;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.net.*;
import java.time.Duration;
import java.time.Instant;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.BlockingQueue;
import java.util.concurrent.LinkedBlockingQueue;
import java.util.concurrent.TimeUnit;

public class Colorfight {
    public int uid = 0;
    public int turn = 0;
    public int max_turn = 0;
    public Double round_time = 0.0;
    public User me = null;
    public Map<Integer, User> users = new HashMap<>();
    public Map<Integer, Object> error = new HashMap<>();
    public String game_id = "0";
    public GameMap game_map = null;

    private WebsocketClientEndpoint clientEndPoint_action;
    private WebsocketClientEndpoint clientEndPoint_info;
    private BlockingQueue<String> action_queue = new LinkedBlockingQueue<>();
    private BlockingQueue<String> info_queue = new LinkedBlockingQueue<>();

    private String parse_url( String url ){
        if ( url.startsWith( "http" ) ) {
            return url.replaceFirst( "http", "wss" );
        }
        else if ( !url.startsWith( "ws" ) ) {
            return  "wss://" + url;
        }
        return url;
    }

    private void _reset() {
        this.uid = 0;
        this.turn = 0;
        this.max_turn = 0;
        this.round_time = 0.0;
        this.me = null;
        this.users = new HashMap<>();
        this.error = new HashMap<>();
        this.game_id = "0";
        this.game_map = null;
        this.action_queue = new LinkedBlockingQueue<>();
        this.info_queue = new LinkedBlockingQueue<>();
    }

    private void _update( JSONObject info ){
        turn = ( (Long) info.get("turn") ).intValue();
        error = (Map<Integer, Object>) info.get("error");
        max_turn = ( (Long) ( (JSONObject)info.get("info") ).get("max_turn") ).intValue();
        round_time = (double) ( (JSONObject) info.get("info") ).get("round_time");

        game_map = new GameMap(
                ( (Long) ( (JSONObject)info.get("info") ).get("width") ).intValue(),
                ( (Long) ( (JSONObject)info.get("info") ).get("height") ).intValue()
        );
        game_map._update_info( (JSONObject) info.get("game_map") );

        for ( Object o:( (Map)info.get("users") ).values() ) {
            JSONObject user =  (JSONObject) o;
            User temp = new User();
            temp._update_info( user );
            ( (Long) user.get("uid") ).intValue();
            users.put( ((Long) user.get("uid") ).intValue(), temp );
        }
        if ( users!=null && users.containsKey(uid) ) {
            me = users.get( uid );
        }
    }

    public void connect() throws URISyntaxException {
        clientEndPoint_action = new WebsocketClientEndpoint(
                new URI( "wss://www.colorfightai.com/gameroom/public/action_channel" )
        );
        clientEndPoint_info = new WebsocketClientEndpoint(
                new URI( "wss://www.colorfightai.com/gameroom/public/game_channel" )
        );

        // add listener
        clientEndPoint_action.addMessageHandler( new WebsocketClientEndpoint.MessageHandler() {
            public void handleMessage( String message ) {
                //System.out.println(message);
                action_queue.add( message );
            }
        } );

        // add listener
        clientEndPoint_info.addMessageHandler( new WebsocketClientEndpoint.MessageHandler() {
            public void handleMessage( String message ) {
                //System.out.println(message);
                info_queue.add( message );
            }
        } );
    }

    public void connect( String room ) throws URISyntaxException {
        String url = "wss://www.colorfightai.com/gameroom/" + room;
        clientEndPoint_action = new WebsocketClientEndpoint( new URI( url+"/action_channel" ) );
        clientEndPoint_info = new WebsocketClientEndpoint( new URI( url+"/game_channel" ) );

        // add listener
        clientEndPoint_action.addMessageHandler( new WebsocketClientEndpoint.MessageHandler() {
            public void handleMessage( String message ) {
                //System.out.println(message);
                action_queue.add( message );
            }
        } );

        // add listener
        clientEndPoint_info.addMessageHandler( new WebsocketClientEndpoint.MessageHandler() {
            public void handleMessage( String message ) {
                //System.out.println(message);
                info_queue.add( message );
            }
        } );
    }

    public void connect_url( String url ) throws URISyntaxException {
        url = parse_url(url);
        clientEndPoint_action = new WebsocketClientEndpoint( new URI( url+"/action_channel" ) );
        clientEndPoint_info = new WebsocketClientEndpoint( new URI( url+"/game_channel" ) );

        // add listener
        clientEndPoint_action.addMessageHandler( new WebsocketClientEndpoint.MessageHandler() {
            public void handleMessage( String message ) {
                //System.out.println(message);
                action_queue.add( message );
            }
        } );

        // add listener
        clientEndPoint_info.addMessageHandler( new WebsocketClientEndpoint.MessageHandler() {
            public void handleMessage( String message ) {
                //System.out.println(message);
                info_queue.add( message );
            }
        } );
    }

    public void disconnect() throws IOException {
        System.out.println( "disconnect!" );
        clientEndPoint_action.userSession.close();
        clientEndPoint_info.userSession.close();
        action_queue.clear();
        info_queue.clear();
        action_queue = null;
        info_queue = null;
        _reset();
    }

    public Boolean update_turn() throws InterruptedException, ParseException {
        JSONParser parser = new JSONParser();
        JSONObject json = (JSONObject) parser.parse( info_queue.take() );
        Instant start = Instant.now();
        while ( true ) {
            while (!info_queue.isEmpty()) {
                json = (JSONObject) parser.parse(info_queue.poll(0, TimeUnit.SECONDS));
            }
            if (this.game_id.equals("0")) {
                this.game_id = (String) ((JSONObject) json.get("info")).get("game_id");
            }

            if (!this.game_id.equals((String) ((JSONObject) json.get("info")).get("game_id"))) {
                return false;
            }

            if ( ( (Long) json.get("turn") ).intValue() != turn ) {
                if ( ( (Long) ( (JSONObject) json.get("info") ).get("game_version") ).intValue() != Constants.GAME_VERSION ) {
                    System.out.println("Your bot is out of date, please update your bot by git pull or download from the website");
                }
                break;
            }
            Thread.sleep( 5 );
        }
        _update( json );
        return true;
    }

    public Boolean update_turn( double timeout ) throws InterruptedException, ParseException {
        JSONParser parser = new JSONParser();
        JSONObject json = (JSONObject) parser.parse( info_queue.take() );
        Instant start = Instant.now();
        while ( true ) {
            while (!info_queue.isEmpty()) {
                json = (JSONObject) parser.parse(info_queue.poll(0, TimeUnit.SECONDS));
            }
            if (this.game_id.equals("0")) {
                this.game_id = (String) ((JSONObject) json.get("info")).get("game_id");
            }

            if ( timeout > 0 && ( Duration.between( start, Instant.now() ).toSeconds() > timeout ) ) {
                return false;
            }

            if (!this.game_id.equals((String) ((JSONObject) json.get("info")).get("game_id"))) {
                return false;
            }

            if ( ( (Long) json.get("turn") ).intValue() != turn ) {
                if ( ( (Long) ( (JSONObject) json.get("info") ).get("game_version") ).intValue() != Constants.GAME_VERSION ) {
                    System.out.println("Your bot is out of date, please update your bot by git pull or download from the website");
                }
                break;
            }
            Thread.sleep( 5 );
        }
        _update( json );
        return true;
    }

    public boolean register( String username, String password ) throws InterruptedException, ParseException {
        clientEndPoint_action.sendMessage(
                "{\"action\":\"register\",\"username\":\""+
                        username+
                        "\",\"password\":\""+
                        password+"\",\"join_key\":\"\"}"
        );
        JSONParser parser = new JSONParser();
        JSONObject json = (JSONObject) parser.parse( action_queue.poll( 2, TimeUnit.SECONDS ) );
        if ( json.get("uid")==null ) {
            System.out.println( json.get("err_msg") );
            return false;
        }
        else {
            uid = ( (Long) json.get("uid") ).intValue();
        }
        return true;
    }

    public boolean register( String username, String password, String join_key ) throws InterruptedException, ParseException {
        clientEndPoint_action.sendMessage(
                "{\"action\":\"register\",\"username\":\""+
                        username+
                        "\",\"password\":\""+
                        password+"\""+
                        "\"join_key\":\""+
                        join_key+"\"}"
        );
        JSONParser parser = new JSONParser();
        JSONObject json = (JSONObject) parser.parse( action_queue.poll( 2, TimeUnit.SECONDS ) );
        if ( json.get("uid")==null ) {
            System.out.println( json.get("err_msg") );
            return false;
        }
        else {
            uid = ( (Long) json.get("uid") ).intValue();
        }
        return true;
    }

    public String attack( Position position, int energy ){
        return Constants.CMD_ATTACK
                + " " + position.x
                + " " + position.y
                + " " + energy;
    }

    public String build( Position position, char building ){
        return Constants.CMD_BUILD
                + " " + position.x
                + " " + position.y
                + " " + building;
    }

    public String upgrade( Position position ){
        return Constants.CMD_UPGRADE
                + " " + position.x
                + " " + position.y;
    }

    public JSONObject send_cmd( ArrayList<String> cmd_list ) throws InterruptedException, ParseException {
        Map<String, Object> msg = new HashMap<>();
        msg.put( "action", "command" );
        msg.put( "cmd_list", cmd_list );
        clientEndPoint_action.sendMessage( new JSONObject(msg).toJSONString() );
        JSONParser parser = new JSONParser();
        // JSONObject is basically the same as a java HashMap.
        return (JSONObject) parser.parse( action_queue.take() );
    }

    public JSONObject get_gameroom_list( String host ) throws IOException, ParseException {
        URL url = new URL(host + "get_gameroom_list");
        HttpURLConnection con = (HttpURLConnection) url.openConnection();
        con.setRequestMethod("GET");
        con.setRequestProperty("Content-Type", "application/json");
        con.setRequestProperty("User-Agent", "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/75.0.3770.90 Safari/537.36");
        System.out.println("\nSending 'GET' request to URL : " + url);
        BufferedReader in = new BufferedReader(
                new InputStreamReader(con.getInputStream()));
        String inputLine;
        StringBuilder response = new StringBuilder();
        while ((inputLine = in.readLine()) != null) {
            response.append(inputLine);
        }
        in.close();
        return (JSONObject) new JSONParser().parse(response.toString());
    }

    public JSONArray get_gameroom_list() throws IOException, ParseException {
        URL url = new URL( "https://www.colorfightai.com/get_gameroom_list" );
        HttpURLConnection con = (HttpURLConnection) url.openConnection();
        con.setRequestMethod("GET");
        con.setRequestProperty("Content-Type", "application/json");
        con.setRequestProperty("User-Agent", "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/75.0.3770.90 Safari/537.36");
        System.out.println("\nSending 'GET' request to URL : " + url);
        BufferedReader in = new BufferedReader(
                new InputStreamReader(con.getInputStream()));
        String inputLine;
        StringBuilder response = new StringBuilder();
        while ((inputLine = in.readLine()) != null) {
            response.append(inputLine);
        }
        in.close();
        return (JSONArray) new JSONParser().parse(response.toString());
    }
}

File: python3/colorfight/colorfight.py
import time
import queue
import json
import urllib.request
import urllib.parse

from .game_map import GameMap
from .user import User
from .position import Position
from .network import Network
from .constants import update_globals, CMD_ATTACK, CMD_BUILD, CMD_UPGRADE, GAME_VERSION

class Colorfight:
    def __init__(self):
        self._reset()

    def _reset(self):
        self.uid = 0
        self.turn = 0
        self.max_turn = 0
        self.round_time = 0
        self.nw = None
        self.me = None
        self.users = {}
        self.error = {}
        self.game_id = 0
        self.game_map = None
        self.info_queue = None
        self.action_queue = None
        self.action_resp_queue = None

    def connect(self, room = 'public', url = None):
        self.info_queue = queue.Queue()
        self.action_queue = queue.Queue()
        self.action_resp_queue = queue.Queue()
        if url == None:
            url = 'https://www.colorfightai.com/gameroom/' + room
        self.nw = Network(self.info_queue, self.action_queue, self.action_resp_queue, url)
        self.nw.setDaemon(True)
        self.nw.start()

    def disconnect(self):
        print("disconnect!")
        self.nw.disconnect()
        del self.info_queue
        del self.action_queue
        del self.action_resp_queue
        self._reset()

    def _update(self, info):
        self.turn = info['turn']
        self.error = info['error']
        self._update_info(info['info'])
        self.game_map = GameMap(self.width, self.height)
        self.game_map._update_info(info['game_map'])
        self.users = {}
        for uid in info['users']:
            user = User()
            user._update_info(info['users'][uid])
            user.cells = {}
            for pos_lst in info['users'][uid]['cells']:
                pos = Position(pos_lst[0], pos_lst[1])
                user.cells[pos] = self.game_map[pos]
            self.users[int(uid)] = user
        if self.uid in self.users:
            self.me = self.users[self.uid]
        else:
            self.me = None

    def _update_info(self, info):
        for field in info:
            setattr(self, field, info[field])
        update_globals(info)

    def update_turn(self, timeout = 0):
        info = self.info_queue.get()
        timer = time.time()
        while True:
            while not self.info_queue.empty():
                info = self.info_queue.get()
            
            if self.game_id == 0:
                self.game_id = info["info"]["game_id"]

            if timeout > 0 and time.time() - timer > timeout:
                return False

            if info["info"]["game_id"] != self.game_id:
                return False

            if info["turn"] != self.turn:
                if info['info']['game_version'] != GAME_VERSION:
                    print("Please update your bot. You can do git pull or download from the website.")
                break
            
            time.sleep(0.005)
                
        self._update(info)
        return True

    def register(self, username, password, join_key = ''):
        self.action_queue.put({'action': 'register', 
                'username': username, 
                'password': password,
                'join_key': join_key
        })
        time.sleep(0.01)
        try:
            result = self.action_resp_queue.get(timeout = 2)
            if "err_msg" in result:
                print(result["err_msg"])
                return False
            else:
                self.uid = int(result['uid'])
                return True
        except Exception as e:
            raise Exception("Failed to register to the game!")

    def attack(self, position, energy):
        '''
            /param position: a Position object for the attacked position
            /param energy: the energy the user uses

            /return: a string representing a command
        '''
        return "{} {} {} {}".format(CMD_ATTACK, position.x, position.y, energy)

    def build(self, position, building):
        '''
            /param position: a Position object for the build position
            /param building: a letter representing the building

            /return: a string representing a command
        '''
        return "{} {} {} {}".format(CMD_BUILD, position.x, position.y, building)
            
    def upgrade(self, position):
        '''
            /param position: a Position object to upgrade

            /return: a string representing a command
        '''
        return "{} {} {}".format(CMD_UPGRADE, position.x, position.y)

    def send_cmd(self, cmd_list):
        msg = {"action": "command", "cmd_list": cmd_list}
        self.action_queue.put(msg)
        result = self.action_resp_queue.get()
        return result

    def get_gameroom_list(self, host = 'https://www.colorfightai.com/'):
        url = urllib.parse.urljoin(host, 'get_gameroom_list')
        # Stupid Cloudflare Checks for Scrapers. I have to disguise my own
        # request to pass it!
        headers = {
            "user-agent": "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/75.0.3770.90 Safari/537.36"
        }
        request = urllib.request.Request(url = url, headers = headers)
        with urllib.request.urlopen(request) as f:
            return json.loads(f.read().decode('utf-8'))
